aten/src/TH/generic/THTensorRandom.cpp

```diff
@@ -51,10 +55,73 @@ void THTensor_(geometric)(THTensor *self, THGenerator *_generator, double p)
   TH_TENSOR_APPLY(real, self, *self_data = (real)THRandom_geometric(_generator, p););
 }
 
+#ifdef TH_BLAS_MKL
+#define BERNOULLI_OMP 10000
+void Bernoulli_generate(const int64_t n, const double p, real *r, THGenerator *_generator)
+{
+  int64_t seed = 17 + THRandom_random(_generator) % 4096;
+
+#ifdef TH_REAL_IS_INT
+  int *tmp = r;
+#else
+  int *tmp = (int*)malloc(n*sizeof(int));
+#endif
+
+#ifdef _OPENMP
+  size_t nthr = omp_get_num_threads();
+  int run_parallel = 0;
+  if ((omp_in_parallel() == 0) && (n >= BERNOULLI_OMP)) {
+    run_parallel = 1;
+  }
+  if (0 == run_parallel) nthr = 1;
+
+#pragma omp parallel num_threads(nthr) firstprivate(nthr)
+  {
+    size_t tid = omp_get_thread_num();
+    int64_t seg_len_tmp = n / nthr;
+    int64_t line_index_offset = tid * seg_len_tmp;
+    int64_t line_seg_len = (tid == nthr - 1)? (n-line_index_offset) : seg_len_tmp;
+#else
+  {
+    int64_t line_index_offset = n;
+    int64_t line_seg_len = 0;
+#endif
+
+    if (line_seg_len > 0) {
+      VSLStreamStatePtr stream;
+      vslNewStream(&stream, VSL_BRNG_MCG31, seed);
+      vslSkipAheadStream(stream, line_index_offset);
+      viRngBernoulli(VSL_RNG_METHOD_BERNOULLI_ICDF, stream, line_seg_len,
+        tmp + line_index_offset, p);
+      vslDeleteStream(&stream);
+#ifndef TH_REAL_IS_INT
+      int64_t i = 0;
+      int64_t line_index_last = line_index_offset + line_seg_len;
+      for(i = line_index_offset; i < line_index_last; i++) {
+        r[i] = (real)tmp[i];
+      }
+#endif
```

Comment:
I think it would be simpler to make this function always fill a `THIntTensor`. Then, in the `bernoulli` function, you'd allocate the tensor if `real` is not an `int`, and use a regular vectorized `copy` from TH to move the data over.

Reply:
It's a nice idea. I will modify the code.